### ISSUE TITLE
Query/Json: improve JSON entity/collection comparisons to null (fails currently for optional dependent with all nullable properties)

### ISSUE THREAD

**Issue by ChesterChuTw** [dotnet/efcore#30589]
I'm using EFCore 7.

I have a ef core model likes below, the column `Address` that allows json string or null:
```csharp
  modelBuilder.Entity<User>(e =>
  {
      e.OwnsOne(p => p.Address, b => b.ToJson());
  });

public partial class User
{
    [StringLength(50)]
    [Unicode(false)]
    public string Name { get; set; } = null!;
    ...
    public UserAddress? Address { get; set; } = null!;
}
```

And I faced two questions:

1. When I executed the query likes below:
```csharp
_labDBContext.Users.Select(x => new Model_User()
{
    Name = x.Name,
     ...
    UserAddress = x.Address
}).ToList();
```
I would get error message `variable '' of type 'IScom_Test_API.Models.DBModels.LabDB.UserAddress' referenced from scope '', but it is not defined`

but if I change to the code as below, it will work as expected:
```csharp
_labDBContext.Users.Select(x => new Model_User()
{
    Name = x.Name,
     ...
    UserAddress = new()
    {
        City = x.Address.City,
        Country = x.Address.Country
    }
}).ToList();
```
I'm not sure if there's any problem or limitation about EFCore's `Select` function or not?

2. And I have another code as below, the same, want to select or filter the IQuerable variable with a json column `Address`:
```csharp
_labDBContext.Users.Where(x => x.Address == null).ToList();
```

It'll also return the errors, 
`Error creating query string: The LINQ expression 'DbSet<User>()
    .Where(u => EF.Property<UserAddress>(u, "Address") == null)' could not be translated. Either rewrite the query in a form that can be translated, or switch to client evaluation explicitly by inserting a call to 'AsEnumerable', 'AsAsyncEnumerable', 'ToList', or 'ToListAsync'. See https://go.microsoft.com/fwlink/?linkid=2101038 for more information..`
    
I think the reasons that caused the two questions are similar, does anyone knows why would it  cause error if I want to manipulate a json column?

**Comment by maumar** [comment]
wrt first problem, if you are also projecting a collection of entities in the same query, it's a duplicate of https://github.com/dotnet/efcore/issues/30266. It's fixed in the current bits so you can unblock yourself by using latest daily build, or use the workaround shown in the original issue (i.e. Include the collection and project entity itself, rather than individual properties/navigations)

**Comment by maumar** [comment]
@ChesterChuTw  wrt the second error, can you provide the full repro code? When I try similar scenario on my end, the query gets translated. Code:

```cs
    public class MyEntity
    {
        public int Id { get; set; }

        public MyJson? Reference { get; set; }

        public List<MyJson>? Collection { get; set; }
    }

    public class MyJson
    {
        public string? Foo { get; set; }
        public int Bar { get; set; }
    }

    public class MyContext : DbContext
    {
        public DbSet<MyEntity>? Entities { get; set; }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.Entity<MyEntity>().Property(x => x.Id).ValueGeneratedNever();
            modelBuilder.Entity<MyEntity>().OwnsOne(x => x.Reference).ToJson();
            modelBuilder.Entity<MyEntity>().OwnsMany(x => x.Collection).ToJson();
        }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            optionsBuilder.UseSqlServer(@"Server=(localdb)\mssqllocaldb;Database=Repro;Trusted_Connection=True;MultipleActiveResultSets=true");
        }
    }


    [ConditionalFact]
    public virtual void Test()
    {
        using (var ctx = new MyContext())
        {
            ctx.Database.EnsureDeleted();
            ctx.Database.EnsureCreated();

            var e1 = new MyEntity
            {
                Id = 1,
                Reference = new MyJson { Bar = 1, Foo = "r1" },
                Collection = new List<MyJson>
                {
                    new MyJson { Bar = 11, Foo = "c11" },
                    new MyJson { Bar = 12, Foo = "c12" },
                }
            };

            var e2 = new MyEntity
            {
                Id = 2,
                Reference = null,
                Collection = null,
            };

            var e3 = new MyEntity
            {
                Id = 3,
                Reference = null,
                Collection = new List<MyJson>(),
            };

            ctx.Entities.AddRange(e1, e2, e3);
        }

        using (var ctx = new MyContext())
        {
            var query = ctx.Entities.Where(x => x.Reference == null).ToList();
        }
    }
```


sql:

```sql
SELECT [e].[Id], [e].[Collection], [e].[Reference]
FROM [Entities] AS [e]
WHERE CAST(JSON_VALUE([e].[Reference], '$.Bar') AS int) IS NULL OR JSON_VALUE([e].[Reference], '$.Foo') IS NULL
```

**Comment by maumar** [comment]
Irrespective, there is room for improvement here, we don't need to go through normal entity to null comparison path, but rather compare JSON column or (in case of nested json entities) JSON fragment storing entire entity to null directly. Something like:

```sql
SELECT [e].[Id], [e].[Collection], [e].[Reference]
FROM [Entities] AS [e]
WHERE [e].[Reference] IS NULL
```

**Comment by ChesterChuTw** [comment]
> wrt first problem, if you are also projecting a collection of entities in the same query, it's a duplicate of #30266. It's fixed in the current bits so you can unblock yourself by using latest daily build, or use the workaround shown in the original issue (i.e. Include the collection and project entity itself, rather than individual properties/navigations)

Thanks for your reply, for the first question, do you mean by using the preview version? 
Because the version I'm using now is the latest version of 7.0.4.

**Comment by ChesterChuTw** [comment]
> Irrespective, there is room for improvement here, we don't need to go through normal entity to null comparison path, but rather compare JSON column or (in case of nested json entities) JSON fragment storing entire entity to null directly. Something like:
> 
> ```sql
> SELECT [e].[Id], [e].[Collection], [e].[Reference]
> FROM [Entities] AS [e]
> WHERE [e].[Reference] IS NULL
> ```

Yeah, there's the exactly problem that I was cofused about, I've tried to query a column in my json column, and it worked as expected.
But unfortunately, when I only want to check if my json column is null or not, it threw the exception.

**Comment by maumar** [comment]
@ChesterChuTw the fix is not in the 7.x, we might patch it in the future for now it didn't meet the bar (only handful of people actually encountered this problem). By daily builds I mean most up to date EF Code (with fixes from yesterday) as described here: https://github.com/dotnet/efcore/blob/main/docs/DailyBuilds.md If you want to avoid risks associated with using such hot bits, your best approach is to use the workaround, which isn't too bad.

**Comment by ChesterChuTw** [comment]
> @ChesterChuTw wrt the second error, can you provide the full repro code? When I try similar scenario on my end, the query gets translated. Code:
> 
> ```cs
>     public class MyEntity
>     {
>         public int Id { get; set; }
> 
>         public MyJson? Reference { get; set; }
> 
>         public List<MyJson>? Collection { get; set; }
>     }
> 
>     public class MyJson
>     {
>         public string? Foo { get; set; }
>         public int Bar { get; set; }
>     }
> 
>     public class MyContext : DbContext
>     {
>         public DbSet<MyEntity>? Entities { get; set; }
> 
>         protected override void OnModelCreating(ModelBuilder modelBuilder)
>         {
>             modelBuilder.Entity<MyEntity>().Property(x => x.Id).ValueGeneratedNever();
>             modelBuilder.Entity<MyEntity>().OwnsOne(x => x.Reference).ToJson();
>             modelBuilder.Entity<MyEntity>().OwnsMany(x => x.Collection).ToJson();
>         }
> 
>         protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
>         {
>             optionsBuilder.UseSqlServer(@"Server=(localdb)\mssqllocaldb;Database=Repro;Trusted_Connection=True;MultipleActiveResultSets=true");
>         }
>     }
> 
> 
>     [ConditionalFact]
>     public virtual void Test()
>     {
>         using (var ctx = new MyContext())
>         {
>             ctx.Database.EnsureDeleted();
>             ctx.Database.EnsureCreated();
> 
>             var e1 = new MyEntity
>             {
>                 Id = 1,
>                 Reference = new MyJson { Bar = 1, Foo = "r1" },
>                 Collection = new List<MyJson>
>                 {
>                     new MyJson { Bar = 11, Foo = "c11" },
>                     new MyJson { Bar = 12, Foo = "c12" },
>                 }
>             };
> 
>             var e2 = new MyEntity
>             {
>                 Id = 2,
>                 Reference = null,
>                 Collection = null,
>             };
> 
>             var e3 = new MyEntity
>             {
>                 Id = 3,
>                 Reference = null,
>                 Collection = new List<MyJson>(),
>             };
> 
>             ctx.Entities.AddRange(e1, e2, e3);
>         }
> 
>         using (var ctx = new MyContext())
>         {
>             var query = ctx.Entities.Where(x => x.Reference == null).ToList();
>         }
>     }
> ```
> 
> sql:
> 
> ```sql
> SELECT [e].[Id], [e].[Collection], [e].[Reference]
> FROM [Entities] AS [e]
> WHERE CAST(JSON_VALUE([e].[Reference], '$.Bar') AS int) IS NULL OR JSON_VALUE([e].[Reference], '$.Foo') IS NULL
> ```

Here's my code and error message in detail:

api endpoint:
```csharp
public async Task<IActionResult> ToJsonTest()
{
      var IQ = _labDBContext.Users.Where(x => x.Address == null);
	return Ok(await IQ.ToListAsync());
}
```

Db Model:
```csharp
public partial class User
{
      [Key]
      public int Id { get; set; }
      ...
      public UserAddress? Address { get; set; }
}
```

DbContext.OnModelCreating:
```csharp
protected override void OnModelCreating(ModelBuilder modelBuilder)
{
	modelBuilder.Entity<User>(e =>
	{
      	     e.OwnsOne(p => p.Address).ToJson();
        });
	...
}
```

Error Message:
```
"The LINQ expression 'DbSet<User>()\r\n    .Where(u => EF.Property<UserAddress>(u, \"Address\") == null)' could not be translated. Either rewrite the query in a form that can be translated, or switch to client evaluation explicitly by inserting a call to 'AsEnumerable', 'AsAsyncEnumerable', 'ToList', or 'ToListAsync'. See https://go.microsoft.com/fwlink/?linkid=2101038 for more information.
at Microsoft.EntityFrameworkCore.Query.QueryableMethodTranslatingExpressionVisitor.<VisitMethodCall>g__CheckTranslated|15_0(ShapedQueryExpression translated, <>c__DisplayClass15_0&)\r\n   at Microsoft.EntityFrameworkCore.Query.QueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)\r\n   at Microsoft.EntityFrameworkCore.Query.RelationalQueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)\r\n   at Microsoft.EntityFrameworkCore.Query.QueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)\r\n   at Microsoft.EntityFrameworkCore.Query.RelationalQueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)\r\n   at Microsoft.EntityFrameworkCore.Query.QueryCompilationContext.CreateQueryExecutor[TResult](Expression query)\r\n   at Microsoft.EntityFrameworkCore.Storage.Database.CompileQuery[TResult](Expression query, Boolean async)\r\n   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileQueryCore[TResult](IDatabase database, Expression query, IModel model, Boolean async)\r\n   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass12_0`1.<ExecuteAsync>b__0()\r\n   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQuery[TResult](Object cacheKey, Func`1 compiler)\r\n   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.ExecuteAsync[TResult](Expression query, CancellationToken cancellationToken)\r\n   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.ExecuteAsync[TResult](Expression expression, CancellationToken cancellationToken)\r\n   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryable`1.GetAsyncEnumerator(CancellationToken cancellationToken)\r\n   at System.Runtime.CompilerServices.ConfiguredCancelableAsyncEnumerable`1.GetAsyncEnumerator()\r\n   at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.ToListAsync[TSource](IQueryable`1 source, CancellationToken cancellationToken)\r\n   at IScom_Test_API.Controllers.EFcoreTestController.ToJsonTest() in D:\\GlassWu\\Visual Studio 2019\\IScom_Test_API\\IScom_Test_API\\Controllers\\EFcoreTestController.cs:line 65\r\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.TaskOfIActionResultExecutor.Execute(ActionContext actionContext, IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)\r\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Logged|12_1(ControllerActionInvoker invoker)\r\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)\r\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)\r\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)\r\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeInnerFilterAsync()\r\n--- End of stack trace from previous location ---\r\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|20_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)\r\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Logged|17_1(ResourceInvoker invoker)\r\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Logged|17_1(ResourceInvoker invoker)\r\n   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)\r\n   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)\r\n   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)\r\n   at IScom_Test_API.Middleware.ExceptionHandleMiddleware.Invoke(HttpContext httpContext) in D:\\User\\Visual Studio 2019\\MyProject\\MyProject\\Middleware\\ExceptionHandleMiddleware.cs:line 45"
```

Please feel free to contact me if you need any further information.
Again, thanks for your helps!

**Comment by ChesterChuTw** [comment]
And I've tried the code that you posted, it would work, but if I changed your model like below, remove the `Bar` column, it would throw the same error as mine

```csharp
    public class MyJson
    {
        public string? Foo { get; set; }
    }
```
Or, as well, if I have json db model like this, it would throw the same errors,
```
public class UserAddress
    {
        public int? Id { get; set; }
        public string? City { get; set; }
    }
```

In summary, if a json model that all the columns are nullable, it would throw errors, and if it contains any other column that is not nullable, it would work as expected.
